eng/tools/GenerateFiles/Directory.Build.targets.in

```diff
@@ -23,37 +23,35 @@
 
   <ItemGroup>
     <!-- Use the same NETCore shared framework as repo built against except when building product code in servicing. -->
-    <KnownFrameworkReference
-        Update="@(KnownFrameworkReference->WithMetadataValue('Identity', 'Microsoft.NETCore.App')->WithMetadataValue('TargetFramework', '${DefaultNetCoreTargetFramework}'))"
```

Comment:
@rainersigwald shouldn't this construct _only_ update the item with the exact metadata described❔ I could swear I tested this out w/ earlier SDKs et cetera and saw it working correctly. Definitely updates all items named `Microsoft.NETCore.App` now.

My preference here would be to avoid repeating the `'%(TargetFramework)' == '${DefaultNetCoreTargetFramework}'` on all metadata updates.

Reply:
/ping @rainersigwald

Reply:
Yeah, I would have expected that to work. @BenVillalobos can you take a look at this?

Reply:
Should I wait to merge this PR @BenVillalobos @rainersigwald❔

Reply:
I also would have expected this to work. I can dig into this tomorrow and get back to you. I wouldn't consider it a blocker unless you're worried about perf / potentially updating it again

Reply:
Let us know if we should update again; we can deal 😃

Reply:
Quick update:

```xml
<Project Sdk="Microsoft.Build.NoTargets/3.1.0">

    <PropertyGroup>
        <TargetFramework>net472</TargetFramework>
    </PropertyGroup>
    
    <ItemGroup>
        <TestItem Include="Foo">
            <TestMetadata>Hello</TestMetadata>
        </TestItem>

        <TestItem Include="Foo">
            <TestMetadata>Hello2</TestMetadata>
        </TestItem>
    </ItemGroup>

    <ItemGroup>
        <TestItem Update="@(TestItem->WithMetadataValue('TestMetadata', 'Hello'))">
            <SomeNewMetadata>test</SomeNewMetadata>
        </TestItem>
    </ItemGroup>

</Project>
```

Modifies both items and adds `SomeNewMetadata` to both. If the second `TestItem` is changed to `TestItem Include="Bar"` then it works as intended.

What does a `KnownFrameworkReference` item look like? Do these items ever overlap in identity/include?

Reply:
Expanding on the previous example, it looks like it still works as it should (assuming there's no overlap)

```xml

<Project Sdk="Microsoft.Build.NoTargets/3.1.0">

    <PropertyGroup>
        <TargetFramework>net472</TargetFramework>
    </PropertyGroup>
    
    <ItemGroup>
        <TestItem Include="Foo">
            <TestMetadata>Hello</TestMetadata>
            <TestMetadata2>Hello!</TestMetadata2>
        </TestItem>
        <TestItem Include="Bar">
            <TestMetadata>NotHello</TestMetadata>
        </TestItem>
        <TestItem Include="Baz">
            <TestMetadata>Hello</TestMetadata>
            <TestMetadata2>Hello</TestMetadata2>
        </TestItem>
    </ItemGroup>

    <ItemGroup>
        <TestItem Update="@(TestItem->WithMetadataValue('TestMetadata', 'Hello')->WithMetadataValue('TestMetadata2', 'Hello'))">
            <SomeNewMetadata>woo!</SomeNewMetadata>
        </TestItem>
    </ItemGroup>

</Project>

```

Results in 

![image](https://user-images.githubusercontent.com/4691428/134242900-69f22bea-24ba-4924-9d66-2b32337d0ddf.png)

Reply:
I think this might be an issue with using ${} instead of $().

```xml
<Project Sdk="Microsoft.Build.NoTargets/3.1.0">

    <PropertyGroup>
        <TargetFramework>net472</TargetFramework>
        <SomeData>Hello</SomeData>
    </PropertyGroup>
    
    <ItemGroup>
        <TestItem Include="Foo">
            <TestMetadata>Hello</TestMetadata>
            <TestMetadata2>Hello!</TestMetadata2>
        </TestItem>
        <TestItem Include="Bar">
            <TestMetadata>NotHello</TestMetadata>
        </TestItem>
        <TestItem Include="Baz">
            <TestMetadata>Hello</TestMetadata>
            <TestMetadata2>Hello</TestMetadata2>
        </TestItem>
    </ItemGroup>

    <ItemGroup>
        <TestItem Update="@(TestItem->WithMetadataValue('TestMetadata', 'Hello')->WithMetadataValue('TestMetadata2', '$(SomeData)'))">
            <SomeNewMetadata>woo!</SomeNewMetadata>
        </TestItem>
    </ItemGroup>

</Project>

```xml

This project properly sets metadata, whereas if I use `'${SomeData}'` it is just set to the literal string `${SomeData}`

Do you see the value of `LatestRuntimeFrameworkVersion` being properly set when using `${}`?

Reply:
1. Please find attached a sample Directory.Build.targets file after we've used the Directory.Build.targets.in template. The final file contains no curly braces.
[Directory.Build.targets.txt](https://github.com/dotnet/aspnetcore/files/7206762/Directory.Build.targets.txt)
2. The `@(KnownFrameworkReference)` item group consists almost entirely of overlapping groups e.g. there are 4 TFM-specific entries for `Microsoft.NETCore.App:
![image](https://user-images.githubusercontent.com/6431421/134259881-1087d870-fd01-4271-8d33-73679c57aa2d.png)
3. When using `->WithMetadataValue('TargetFramework', '${DefaultNetCoreTargetFramework}'))` (actually `->WithMetadataValue('TargetFramework', 'net6.0'))`), the result was all four `Microsoft.NETCore.App` items were updated.

Reply:
@dougbu do you know what version of the sdk you saw this working prior?

I've filed an issue for it here: https://github.com/dotnet/msbuild/issues/6880

Reply:
I switched to use the double `WithMetadataValue(...)` in `Includes` way back in 5fd1db26257, December 2020. My old friend `git switch --detach 5fd1db26257` showed me we were using the `6.0.100-alpha.1.20523.3` back then.

Reply:
Oh my. I remember doing a lot of .binlog checking while working on that. But, I just changed the indentation of the relevant Directory.Build.targets.in bits. Looking back&hellip;

Reply:
Introduced in 219ecd688012, October 2020. SDK at _that_ time was also `6.0.100-alpha.1.20523.3` (surprisingly)